library/alloc/src/vec.rs

```diff
@@ -1257,6 +1257,10 @@ impl<T> Vec<T> {
     #[inline]
     #[stable(feature = "append", since = "1.4.0")]
     pub fn append(&mut self, other: &mut Self) {
+        if self.len == 0 && self.capacity() < other.len {
```

Comment:
Does ` && self.capacity() < other.len` matter here? Why not just swap without checking the capacity?

Reply:
Vec guarantees it will never shrink the capacity: https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#guarantees

> Vec will never automatically shrink itself, even if completely empty. This ensures no unnecessary allocations or deallocations occur.

Possibly we could use `self.reserve(other.capacity)` but I'm not sure whether that would be a speed up or not, I'd want a separate perf run for it (after this gets merged).

Reply:
> Vec guarantees it will never shrink the capacity: https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#guarantees

Doesn't this shrink the capacity of the other `Vec`?

Reply:
> Vec guarantees it will never shrink the capacity: https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#guarantees

But doesn't this shrink the capacity of the other `Vec` with `mem::swap`?

Reply:
If the LHS capacity were larger then it might be that someone did a

```rust
let v = Vec::with_capacity(...);
v.append(other);
```

to preallocated, e.g. because `append` called in a loop. It would be surprising if then a smaller vector would get swapped in, negating the preallocation.

Reply:
But wouldn't it be surprising for them if suddenly the other vec just shrink automatically? This seemed to broke vec guarantees.

Reply:
Well, I can always claim that there is no vec that's shrinking, it's merely two vecs changing places.
But yeah that's bending words and getting additional opinions on this might be good.

Reply:
I don't know who is in the alloc team but maybe @Dylan-DPC knows?

Reply:
I don't think it's expected from the current guarantees, that a passed vector have a smaller capacity than before. However I'm not an expert on this and the libs-team should look into this. 

cc @Amanieu

Reply:
Note the second part:

> Vec will never automatically shrink itself, even if completely empty. **This ensures no unnecessary allocations or deallocations occur.**

The swap avoids growing (deallocation + allocation) on the LHS. So it's achieving the underlying goal. Perhaps the guarantee itself is suboptimally phrased.

Reply:
This could be rephrased to something like 

> `Vec` will never automatically shrink itself, if allocations or deallocations would be needed.

Reply:
One scenario where the swap approach would be counter-productive is the following:

```rust
let mut lhs = Vec::new();
let mut rhs = Vec::with_capacity(1000_0000);

rhs.resize_with(10, Default::default);
lhs.append(&mut rhs); // A
rhs.resize_with(1000_0000, Default::default); // B

// ... do additional work with rhs
```

Step A would now move the allocation to lhs and thus B would incur a second big allocation which wouldn't have happened without this change.

I think I can avoid this issue by only performing the swap when the moved allocation matches Vec's growth strategy, i.e. when it is not excessively large. This should work out as an improvement because all we do is trade a potential allocation _of similar size_ in the future to avoid a certain allocation now.
With the current code the issue is that the allocation in the future could be much larger than the allocation we're saving now.

Reply:
@the8472 How about we move this part for another pull request for discussion, I think the rest looks good.

Reply:
Ok, will do.

Reply:
Created #77538 for separate discussion.